34 clean code base

GravityDarkLab opened the pull request:
## Description:

This PR significantly enhances our codebase's structure, cleanliness, and functionality. We've addressed key areas of improvement, ensuring a robust foundation for future development.

## Related Issues:
- resolves #44 
- resolves #34 
- resolves #71 
- resolves #70 

## Explanation of changes:

1. Improved Code Structure and Cleanliness:
2. Some Common components have been developed as reusable utilities, enhancing code reusability and maintainability.
3. Introduced Logging specifically in the `handlers`
4. Utilised flutter_secure_storage for secure JWT storage.
5. Incorporated Jakob's feedbacks

GravityDarkLab committed:
Add config directory for better structure

GravityDarkLab committed:
Refactor ModelGenerator

GravityDarkLab committed:
Add Logging in handlers

GravityDarkLab committed:
Add Logging in handlers

GravityDarkLab committed:
Add secure storage for the token

GravityDarkLab committed:
removed Bootstrap

GravityDarkLab committed:
Add secure storage for the token

carlobortolan committed:
Update `main.dart` to use `StreamBuilder` instead of `StreamProvider`

GravityDarkLab committed:
Refactor  UserViewModel

carlobortolan committed:
Replace old models with proto models

GravityDarkLab committed:
Fix linting

GravityDarkLab committed:
Improve code structure

GravityDarkLab committed:
Add Theme

GravityDarkLab committed:
Refactor 'internal_login_screen.dart'

GravityDarkLab committed:
Refactor

GravityDarkLab committed:
Refactor: Remove unused TextEditingController instances in WelcomeScreen

- Eliminated declarations of usernameController and passwordController in WelcomeScreen as they were not used.
- Ensured clean and optimized code by removing unnecessary Riverpod provider references.
- Improved overall code readability and maintainability.

GravityDarkLab committed:
Optimize CourseOverview to Prevent Unnecessary Re-renders

- Resolved issue causing re-renders of AppBar and BottomNavigationBar on tab changes.
- Implemented Column layout to display 'My Courses' and 'Public Courses' sections together, ensuring consistent visibility.
- Added spacing between course sections for better visual clarity and user experience.
- Enhanced layout performance and overall responsiveness of the CourseOverview screen.

GravityDarkLab committed:
Add code documentation for course_overview and course_section

GravityDarkLab committed:
Refactor Course Sections for Reusability and Clarity

- Introduced CourseContentScreen to centralize the display logic of course sections.
- Refactored PinnedCourses and DownloadsScreen to use CourseContentScreen, enhancing code reusability.
- Added in-code documentation for clarity and future maintainability.

GravityDarkLab committed:
Fix INT64 type

GravityDarkLab committed:
Fix Linting

GravityDarkLab committed:
Fix colors on TabBar not changing correctly

Anishyou committed:
Added list view builder in course_view

Anishyou committed:
changed list view to list view builder in course_view and addressed some trailing commas issue

Anishyou committed:
added material design switch in setting

Anishyou committed:
deleteing course outline

Anishyou committed:
Added base_view for app bar and bottom navigatoe

Anishyou committed:
Added base_view in course_Screen

Anishyou committed:
added onViewAll functionality

Anishyou committed:
no re rendering when click on same page

GravityDarkLab committed:
Improve the usage of the user Model view.
- Added loading state when clicked on login
    - `handleSSOLogin()`
    - `handleSSOLogin()`
    - Added `BehaviorSubject<bool> isLoading = BehaviorSubject.seeded(false);`

GravityDarkLab committed:
Refactor `course_card_view.dart` and added documentation for `base_view.dart`, `course_card_view.dart`, `course_overview_view.dart`

GravityDarkLab committed:
Refactor `ssoAuth` and added documentation

GravityDarkLab committed:
Refactor `SettingsScreen` and added documentation to `internal_login_view` and `welcome_screen_view`

GravityDarkLab committed:
Fix in `ssoAuth` when ` Navigator.pushNamed(context, '/home');` called to set the loading state of login to false.

GravityDarkLab committed:
Fix in `ssoAuth` when ` Navigator.pushNamed(context, '/home');` called to set the loading state of login to false.

carlobortolan commented on lib/config/api_config.dart:
```diff
@@ -0,0 +1,4 @@
+class ApiConfig {
```
What's the idea behind having an ApiConfig if all the API routes are already defined in `Routes.dart`?

GravityDarkLab replied:
Yes make sense.
Done ✅

carlobortolan commented on lib/views/utils/routes.dart:
I think that it would be more appropriate if `routes.dart` was moved to `lib/config` as it doesn't have much to do with the `lib/views/utils` directory in which it currently is.

GravityDarkLab replied:
Yes make sense.
Done ✅

carlobortolan commented on lib/config/app_config.dart:
```diff
@@ -0,0 +1,5 @@
+class AppConfig {
```
Duplicate/Unused `AppConfig` class; see comment in `lib/views/utils/routes.dart`.

GravityDarkLab replied:
Yes make sense.
Done ✅

GravityDarkLab committed:
Rename `routes.dart` to `app_config.dart` and moved to the `config` package

jakobkoerber commented on lib/views/login_view/internal_login_view.dart:
```diff
@@ -0,0 +1,163 @@
+    );
+  }
+
+  Future<void> handleBasicLogin(
```
Putting Logic in View is not a best practice, especially if it mainly interacts with one ViewModel -> maybe move it there? The TextFieldControllers could also be there.

GravityDarkLab replied:
done ✅

jakobkoerber commented on lib/views/course_view/downloaded_pinned_courses_view/content_view.dart:
```diff
@@ -0,0 +1,31 @@
+///   [videoCards] - A list of VideoCard widgets to be displayed under the title.
+class CourseContentScreen extends ConsumerWidget {
+  final String title;
+  final List<Widget> videoCards;
```
Can it be a different Widget? Or is it almost the same?

jakobkoerber replied:
To clarify: I meant the widgets in `List<Widget> videoCards`

GravityDarkLab replied:
Ah got it, yes true in this case it will be a `VideoCard` Widget.
Updated accordingly ✅

jakobkoerber commented on lib/views/course_view/components/courselist_screen.dart:
```diff
@@ -0,0 +1,47 @@
+/// dynamically generates a horizontal list of courses.
+class CourseListScreen extends ConsumerWidget {
+  final String title;
+  final List<Widget> videoCards;
```
Can it be a different Widget? Or is it almost the same?

jakobkoerber replied:
To clarify: I meant the widgets in `List<Widget> videoCards`

jakobkoerber commented on lib/views/course_view/components/course_screen.dart:
```diff
@@ -0,0 +1,74 @@
+/// course lists.
+class CoursesScreen extends StatelessWidget {
+  final String title;
+  final List<Widget> courseCards;
```
Can it be a different Widget? Or is it almost the same?

jakobkoerber replied:
Sorry for the confusion: I think it would be much easier and more efficient to pass a list of data and use ListView.builder to create a list of widgets for each data entry.

GravityDarkLab replied:
(see above)

jakobkoerber commented on lib/views/utils/globals.dart:
```diff
@@ -0,0 +1,25 @@
+final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();
+
+// Provider for managing the lifecycle of the username text controller
+final usernameControllerProvider =
```
Which views use this controller? Or would a ViewModel containing both controllers make more sense here?

GravityDarkLab replied:
Found them being used in WelcomeScreen but had no affects.
I removed them

jakobkoerber commented on lib/views/course_view/components/course_overview_section.dart:
```diff
@@ -0,0 +1,123 @@
+/// different titles, courses and onViewAll actions.
+class CourseSection extends StatelessWidget {
+  final String sectionTitle;
+  final List<Widget>? courses;
```
Can it be a different Widget? Or is it almost the same?

jakobkoerber reviewed (changes requested):
lgtm! **One major issue**: the navigation bar is still consisting of 5 different instances and re-renders completely everytime the index changes. I highly suggest moving to one nav bar.  The rest is just some minor feedback/suggestions. One question: is the login via TUM supposed to work or do I need the backend running locally for that?

carlobortolan commented:
> lgtm! **One major issue**: the navigation bar is still consisting of 5 different instances and re-renders completely everytime the index changes. I highly suggest moving to one nav bar. The rest is just some minor feedback/suggestions. One question: is the login via TUM supposed to work or do I need the backend running locally for that?

Regarding the TUM Login: You need the [current API](https://github.com/tum-Dev/gocast/tree/IPraktikum-36-user-endpoints) running locally. Note, that currently only the internal login will let you access the app; the TUM shibboleth login will be working once the new gRPC API is deployed. (Currently, you can log in using TUM shibboleth, but you'll get an `Authentication failed` error, as the jwt you received is signed by live.rbg.tum.de and the app makes all other API calls (currently) to localhost for which the jwt is invalid)

carlobortolan reviewed (commented):
Looks good :+1: 
Only suggestion is to use the `isLoading` provided by the `UserState` instead of redefining it in the UserViewModel and fix the re-rendering of the views when chaning tabs (see Jakob's comments)

GravityDarkLab committed:
**Refactored** `UserViewModel` and Login Handling**: Moved login logic and `TextEditingController`s into `UserViewModel`

GravityDarkLab committed:
## Refactor: Integrate StateNotifier in UserViewModel
- Refactored `UserViewModel` to extend `StateNotifier<UserState>`.
- Eliminated separate `isLoading` BehaviorSubject in `UserViewModel`. All state changes, including loading states, are now handled directly through the unified `UserState` managed by the notifier.

GravityDarkLab commented on lib/view_models/UserViewModel.dart:
```diff
@@ -0,0 +1,82 @@
+class UserViewModel {
+  final Logger _logger = Logger();
+
+  BehaviorSubject<bool> isLoading = BehaviorSubject.seeded(false);
```
done ✅

GravityDarkLab commented on lib/view_models/UserViewModel.dart:
```diff
@@ -0,0 +1,82 @@
+    current.close();
+  }
+
+  void clearError() {
```
done ✅

GravityDarkLab committed:
## Refactor: Courses views to have `List<VideoCard>` or `List<VideoCard>

GravityDarkLab committed:
## Refactor: Courses views to have to make better use of `List.builder` and added a temporary `CourseModel` in course_model.dart that will be replaced once API courses retrieving works.

jakobkoerber reviewed (approved):
lgtm!
